keps/sig-cloud-provider/2025-extend-konnectivity-for-both-directions/README.md

```diff
@@ -0,0 +1,694 @@
+* **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+* **Konnectivity Server** The proxy server which runs in the master network. It has a secure channel established to the node network. It could work on either a HTTP Connect mechanism or gRPC. If the former it would expose a gRPC interface to KAS to provide connectivity service. If the latter it would use standard HTTP Connect. Formerly known as the Network Proxy Server.
+* **Konnectivity Agent** A proxy agent which runs in the node network for establishing the tunnel. Formerly known as the Network Proxy Agent.
+* **Flat Network** A network where traffic can be successfully routed using IP. Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Currently the Konnectivity Server is accepting requests from the KAS either with the gRPC or the HTTP Connect interfaces and is taking care of forwarding the traffic to the Konnectivity Agent using the previously established connections (initiated by the agents).
+
+In order to enable traffic from Kubelets and pods running on Node Network, the Konnectivity Agents have to expose an endpoint that will be listening on a specific port and forward the traffic to the KAS on the Master Network. As opposed to the traffic flowing from the Master Network to the Node Network, the Konnectivity Agent should act transparently: From a Kubelets or pods standpoint, the Konnectivity Agent should be the final destination instead of acting as a proxy.
+
+The reason why we do not intend to use the same strategy used in the other direction is that we do not have control over the clients using the Kubernetes default service to communicate with the KAS.
+
+In order to enable the feature the following configurations are needed:
+
+* On the Konnectivity Agent, enable the feature gate, define the listening port and addess/port of the KAS.
+* On Konnectivity server, enable the feature gate and allow the address/port to reach the KAS.
+* On KAS, configure the IP address and port that will be used by the agent with `advertise-address` and the `secure-port` flags.
+
+### Traffic Flow
+
+```
+client =TCP=> (:6443) agent GRPC=> server =TCP=> KAS(:6443)
```

Comment:
I'm fine with port 6443 being the default but I would suggest that be configurable.

Reply:
In a HA setup there could be multiple servers that the agent is connected to. Do we have a reason to care where the traffic goes? (Eg. matching failure zone?) Or do we just pick a random server?

Reply:
it is actually, I put 6443 just as an example here as it is the default used by KAS (if no `secure_port` flag is specified)

Reply:
> In a HA setup there could be multiple servers that the agent is connected to. Do we have a reason to care where the traffic goes? (Eg. matching failure zone?) Or do we just pick a random server?

@cheftako I would say that we pick a random server, but if we can think about evolving this later in case of need.